src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+                    break
+                }
+                // TODO: Is there a way to append to this ReadOnlyArray so it doesn't constantly pop focus back to the top?
+                // on reset, first iteration should clear the quickPick's items.
+                // should handle cases where call latencies are stable
+                if (scopeIterationTime === this.activeIterationTime) {
+                    this.quickPick.items = reset ? newItems.value : this.quickPick.items.concat(newItems.value)
+                    // nothing else to reset from here on out
+                    reset = false
+                    this.isDone = newItems.done ?? false
+                } else {
+                    // another newer iteration cycle is in-flight. Break without mutating state.
+                    break
+                }
+            } catch (e) {
+                // append error node
+                // clicking error node should either go backwards (return undefined) or refresh
+                // maybe have one of each?
+                // give quickpick an error message
+                // we should not blow away the existing items, they should still be viable
+                const err = e as Error
+                this.quickPick.items = [
+                    ...this.quickPick.items,
+                    {
+                        ...IteratingQuickPickController.ERROR_ITEM,
+                        detail: err.message,
+                    },
+                ]
+                this.isDone = true
+            }
+        }
+
+        // wrap up, but only if the fucntion call is the currently active one.
+        if (scopeIterationTime === this.activeIterationTime) {
+            // no items in response
+            if (this.isDone && this.quickPick.items.length === 0) {
+                this.quickPick.items = [IteratingQuickPickController.NO_ITEMS_ITEM]
+            }
+
+            // disable loading bar
+            this.quickPick.busy = false
+        }
+    }
+}
+
+/**
+ * Represents an iterator that tranforms another iterator into an array of QuickPickItems.
+ * Additionally, has a reset functionality to reset the iterator to its initial state.
+ */
+export class IteratingQuickPickPopulator<TResponse> {
+    private iterator: AsyncIterator<TResponse>
+
+    /**
+     * @param iteratorFactory Function that returns an iterator, with all default state values set. E.g. `collectionUtils.getPaginatedAwsCallIter`
+     * @param transform Function which transforms a response from the iterator into an array of `vscode.QuickPickItem`s.
+     */
+    public constructor(
+        private readonly iteratorFactory: () => AsyncIterator<TResponse>,
+        private readonly transform: (response: TResponse) => vscode.QuickPickItem[]
+    ) {
+        this.iterator = this.iteratorFactory()
+    }
+
+    /**
+     * Resets the iterator to the default state provided by the iteratorFactory.
+     */
+    public reset(): void {
```

Comment:
Rather than having an explicit `reset` method, couldn't you just have `getPickIterator` always create a new iterator?

It doesn't make that much sense for `getPickIterator` to be called multiple times just to return multiple iterators over the same internal iterator. This would lead to unexpected behavior seen by the callers. Their iterator would essentially skip values if another iterator iterated past where they were looking.

Also, if you make this always return a new iterator, you don't actually need a full class at that point since the state is entirely within the function. You could just have a function that returns a wrapper for the iterator.

In fact, you don't necessarily need anything at all. You could just have users call the iterator factory directly whenever they want to create a new item iterator.

Reply:
I think that might work, I'm always paranoid around managing an iterator's state.

Reply:
I think it's best to take steps to avoid sharing iterators to avoid unexpected behavior. It's kind of like sharing a for loop with another piece of code.